libsolutil/LazyInit.h

```diff
@@ -0,0 +1,120 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <libsolutil/Assertions.h>
+#include <libsolutil/Exceptions.h>
+
+#include <memory>
+#include <optional>
+#include <type_traits>
+#include <utility>
+
+namespace solidity::util
+{
+
+DEV_SIMPLE_EXCEPTION(BadLazyInitAccess);
+
+/**
+ * Represents a lazy-initialized value.
+ * @tparam T the type of the lazy-initialized object; may not be a function, reference, array, or void type; may be const-qualified.
+ */
+template<typename T>
+class LazyInit
+{
+public:
+	using value_type = T;
+
+	static_assert(std::is_object_v<value_type>, "Function, reference, and void types are not supported");
+	static_assert(!std::is_array_v<value_type>, "Array types are not supported.");
+	static_assert(!std::is_volatile_v<value_type>, "Volatile-qualified types are not supported.");
+
+	LazyInit() = default;
+
+	LazyInit(LazyInit const&) = delete;
+
+	// Move constructor must be overridden to ensure that moved-from object is left empty.
+	constexpr LazyInit(LazyInit&& _other) noexcept
+	: m_value(std::move(_other.m_value))
+	{
+		_other.m_value.reset();
+	}
+
+	// Copy and swap idiom.
+	LazyInit& operator=(LazyInit _other) noexcept
+	{
+		this->m_value.swap(_other.m_value);
+	}
+
+	template<typename F>
+	value_type& init(F&& _fun)
+	{
+		doInit(std::forward<F>(_fun));
+		return value();
+	}
+
+	template<typename F>
+	value_type const& init(F&& _fun) const
+	{
+		doInit(std::forward<F>(_fun));
+		return value();
+	}
+
+	constexpr value_type& value()
+	{
+		assertInitialized();
+		return m_value.value();
+	}
+
+	constexpr value_type const& value() const
+	{
+		assertInitialized();
+		return m_value.value();
+	}
+
+	value_type& operator*() { return value(); }
```

Comment:
Do we need `value` if we have this?

Reply:
I think my goal was to mimic std::optional as closely as possible, but I guess that's unnecessary. I'll remove `value`.